File: core/src/test/scala/diplomacy/variant/standard/rule/BuildOrderGeneratorSpec.scala
package diplomacy.variant.standard.rule

import diplomacy.UnitSpec
import diplomacy.board._
import diplomacy.variant.standard.map
import diplomacy.variant.standard.map.Power
import diplomacy.variant.standard.map.Keywords._
import diplomacy.variant.standard.board.{ Turn => T }
import diplomacy.variant.standard.board.Keywords._
import diplomacy.variant.standard.board.Implicits._

class BuildOrderGeneratorSpec extends UnitSpec {
  val generator = new BuildOrderGenerator[T, map.Power]

  "A movement-order-generator" when {
    "there are no powers that can build or disband units" should {
      "use an empty set to skip the build phase." in {
        val board =
          new Board[T, map.Power](
            map.map, 1901.Spring - Build,
            Set(), Map(), Map(), Map()
          )
        val $ = StandardRuleOrderHelper(board)
        import $._

        generator.ordersToSkipPhase(board) should be(Some(Set()))
      }
    }

    "all disband orders are automatically decidable" should {
      "use disband orders to skip the build phase." in {
        val board = {
          new Board[T, map.Power](
            map.map, 1901.Spring - Build,
            Set(
              DiplomacyUnit(Germany, Army, Ruh)
            ), Map(), Map(), Map()
          )
        }
        val $ = StandardRuleOrderHelper(board)
        import $._

        generator.ordersToSkipPhase(board) should be(Some(Set(A(Ruh).disband())))
      }
    }


    "otherwise" can {
      "not skip the build phase." in {
        val board =
          new Board[T, map.Power](
            map.map,
            1901.Autumn - Build,
            Set(
              DiplomacyUnit(Germany, Army, Mun),
              DiplomacyUnit(Italy, Fleet, Nap)
            ),
            Map(
              Rom.province -> Italy, Nap.province -> Italy,
              Mar.province -> France
            ),
            Map(), Map()
          )
        generator.ordersToSkipPhase(board) should be(None)
      }
    }
  }
}

File: test/variant/standard/rule/build-order-generator-spec.js
'use strict'

const chai = require('chai')

const { Board, Unit } = require('../../../../lib/board/package')
const { map, location, Power } = require('../../../../lib/variant/standard/map/package')
const { Army } = require('../../../../lib/variant/standard/rule/military-branch')
const { Disband } = require('../../../../lib/variant/standard/rule/order')
const ProvinceStatus = require('../../../../lib/variant/standard/rule/province-status')
const BuildOrderGenerator = require('../../../../lib/variant/standard/rule/build-order-generator')

chai.should()

const $ = location
const generator = new BuildOrderGenerator()

describe('A MovementOrderGenerator', () => {
  describe('when there are no powers that can build or disband units', () => {
    it('uses an empty to skip the build phase.', () => {
      const board = new Board(map, null, [], [], [])
      generator.ordersToSkipPhase(board).should.deep.equal(new Set())
    })
  })
  describe('when all disband orders are automatically decidable', () => {
    it('uses disband order to skip the build phase.', () => {
      const board = new Board(map, null, [new Unit(Army, $.Ruh, Power.Germany)], [], [])
      generator.ordersToSkipPhase(board).should.deep.equal(new Set([
        new Disband(new Unit(Army, $.Ruh, Power.Germany))
      ]))
    })
  })
  describe('otherwise', () => {
    it('does not skip the build phase', () => {
      const board = new Board(
        map, null,
        [new Unit(Army, $.Ruh, Power.Germany)],
        [],
        [
          [$.Mun.province, new ProvinceStatus(Power.Germany, false)],
          [$.Ber.province, new ProvinceStatus(Power.Germany, false)]
        ]
      );
      (generator.ordersToSkipPhase(board) == null).should.deep.equal(true)
    })
  })
})
